design/encryption-data-at-rest.md

```diff
@@ -0,0 +1,233 @@
+* A FDB cluster host as part of processing request would do the following:
+    1. StorageServer would read desired data blocks from the persistent storage.
+    2. Regenerate the encryption key required to decrypt the data.
+    3. Decrypt data and pass results as plaintext to the caller.
+
+
+Below diagram depicts the end-to-end encryption workflow detailing various modules involved and their interactions. The following section discusses detailed design for involved components.
+
+```
+                                          _______________________________________________________
+                                         |                FDB CLUSER HOST                        |
+                                         |                                                       |
+       _____________________             |   ________________________      _________________     |
+      |                     | (proprietary) |                        |    |                 |    |
+      |                     |<---------- |--|      KMS CONNECTOR     |    |  COMMIT PROXIES |    |
+      |  ENCRYPTION KEY     |            |  |                        |    |                 |    |
+      | MANAGEMENT SOLUTION |            |  |(non FDB - proprietary) |    |                 |    |
+      |                     |            |  |________________________|    |_________________|    |
+      |                     |            |              ^                        |               |
+      |_____________________|            |              | (REST API)             | (Encrypt      |
+                                         |              |                        V  Mutation)    |
+                                         |      _________________________________________        |            __________________
+                                         |     |                                         |       |           |                  |
+                                         |     |           ENCRYPT KEYPROXY SERVER       |<------|-----------|                  |
+                                         |     |_________________________________________|       |           |                  |
+                                         |                          |                            |           |    BACKUP FILES  |
+                                         |                          |  (Encrypt Node)            |           |                  |
+                                         |                          V                            |           |                  |
+                                         |      _________________________________________        |           |  (Encrypt file)  |
+                                         |     |                                         |<------|-----------|                  |
+                                         |     |      REDWOOD STORAGE SERVER             |       |           |__________________|
```

Comment:
should we discuss what features are required for a different storage engine to use this?

Reply:
As per offline discussion, added a disclaimer regarding Redwood being the only SS integration supported by the design. Thanks.